File: lib/f1_types/packet_6_car_telemetry_data.py
# MIT License
#
# Copyright (c) [2024] [Ashwin Natarajan]
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import struct
from typing import Dict, Any, List
from .common import _validate_parse_fixed_segments
from .header import PacketHeader
from .base_pkt import F1PacketBase, F1SubPacketBase

# --------------------- CLASS DEFINITIONS --------------------------------------

class CarTelemetryData(F1SubPacketBase):
    """
    A class representing telemetry data for a single car in a racing simulation.

    Attributes:
        m_speed (int): Speed of the car in kilometers per hour.
        m_throttle (float): Amount of throttle applied (0.0 to 1.0).
        m_steer (float): Steering value (-1.0 (full lock left) to 1.0 (full lock right)).
        m_brake (float): Amount of brake applied (0.0 to 1.0).
        m_clutch (int): Amount of clutch applied (0 to 100).
        m_gear (int): Gear selected (1-8, N=0, R=-1).
        m_engineRPM (int): Engine RPM.
        m_drs (int): DRS (Drag Reduction System) state (0 = off, 1 = on).
        m_revLightsPercent (int): Rev lights indicator (percentage).
        m_revLightsBitValue (int): Rev lights represented as a bitfield.
        m_brakesTemperature (List[int]): List of brake temperatures (celsius) for each wheel.
        m_tyresSurfaceTemperature (List[int]): List of surface temperatures (celsius) for each tyre.
        m_tyresInnerTemperature (List[int]): List of inner temperatures (celsius) for each tyre.
        m_engineTemperature (int): Engine temperature (celsius).
        m_tyresPressure (List[float]): List of tyre pressures (PSI) for each tyre.
        m_surfaceType (List[int]): List of surface types for each wheel, see appendices.

            Note:
                The length of each list attribute should be 4, corresponding to the four wheels of the car.
    """

    COMPILED_PACKET_STRUCT = struct.Struct("<"
        "H" # uint16    m_speed;                    // Speed of car in kilometres per hour
        "f" # float     m_throttle;                 // Amount of throttle applied (0.0 to 1.0)
        "f" # float     m_steer;                    // Steering (-1.0 (full lock left) to 1.0 (full lock right))
        "f" # float     m_brake;                    // Amount of brake applied (0.0 to 1.0)
        "B" # uint8     m_clutch;                   // Amount of clutch applied (0 to 100)
        "b" # int8      m_gear;                     // Gear selected (1-8, N=0, R=-1)
        "H" # uint16    m_engineRPM;                // Engine RPM
        "B" # uint8     m_drs;                      // 0 = off, 1 = on
        "B" # uint8     m_revLightsPercent;         // Rev lights indicator (percentage)
        "H" # uint16    m_revLightsBitValue;        // Rev lights (bit 0 = leftmost LED, bit 14 = rightmost LED)
        "4H" # uint16    m_brakesTemperature[4];     // Brakes temperature (celsius)
        "4B" # uint8     m_tyresSurfaceTemperature[4]; // Tyres surface temperature (celsius)
        "4B" # uint8     m_tyresInnerTemperature[4]; // Tyres inner temperature (celsius)
        "H" # uint16    m_engineTemperature;        // Engine temperature (celsius)
        "4f" # float     m_tyresPressure[4];         // Tyres pressure (PSI)
        "4B" # uint8     m_surfaceType[4];           // Driving surface, see appendices
    )
    PACKET_LEN = COMPILED_PACKET_STRUCT.size

    __slots__ = (
        "m_speed",
        "m_throttle",
        "m_steer",
        "m_brake",
        "m_clutch",
        "m_gear",
        "m_engineRPM",
        "m_drs",
        "m_revLightsPercent",
        "m_revLightsBitValue",
        "m_brakesTemperature",
        "m_tyresSurfaceTemperature",
        "m_tyresInnerTemperature",
        "m_engineTemperature",
        "m_tyresPressure",
        "m_surfaceType",
    )

    def __init__(self, data) -> None:
        """
        Initializes a CarTelemetryData object by unpacking the provided binary data.

        Parameters:
            data (bytes): Binary data to be unpacked.

        Raises:
            struct.error: If the binary data does not match the expected format.
        """
        self.m_brakesTemperature = [0] * 4
        self.m_tyresSurfaceTemperature = [0] * 4
        self.m_tyresInnerTemperature = [0] * 4
        self.m_tyresPressure = [0] * 4
        self.m_surfaceType = [0] * 4

        (
            self.m_speed,
            self.m_throttle,
            self.m_steer,
            self.m_brake,
            self.m_clutch,
            self.m_gear,
            self.m_engineRPM,
            self.m_drs,
            self.m_revLightsPercent,
            self.m_revLightsBitValue,
            self.m_brakesTemperature[0],
            self.m_brakesTemperature[1],
            self.m_brakesTemperature[2],
            self.m_brakesTemperature[3],
            self.m_tyresSurfaceTemperature[0],
            self.m_tyresSurfaceTemperature[1],
            self.m_tyresSurfaceTemperature[2],
            self.m_tyresSurfaceTemperature[3],
            self.m_tyresInnerTemperature[0],
            self.m_tyresInnerTemperature[1],
            self.m_tyresInnerTemperature[2],
            self.m_tyresInnerTemperature[3],
            self.m_engineTemperature,
            self.m_tyresPressure[0],
            self.m_tyresPressure[1],
            self.m_tyresPressure[2],
            self.m_tyresPressure[3],
            self.m_surfaceType[0],
            self.m_surfaceType[1],
            self.m_surfaceType[2],
            self.m_surfaceType[3],
        ) = self.COMPILED_PACKET_STRUCT.unpack(data)

        self.m_drs = bool(self.m_drs)

    def __str__(self) -> str:
        """
        Returns a string representation of the CarTelemetryData object.

        Returns:
            str: String representation of the object.
        """
        return (
            f"CarTelemetryData("
            f"Speed: {self.m_speed}, "
            f"Throttle: {self.m_throttle}, "
            f"Steer: {self.m_steer}, "
            f"Brake: {self.m_brake}, "
            f"Clutch: {self.m_clutch}, "
            f"Gear: {self.m_gear}, "
            f"Engine RPM: {self.m_engineRPM}, "
            f"DRS: {self.m_drs}, "
            f"Rev Lights Percent: {self.m_revLightsPercent}, "
            f"Brakes Temperature: {str(self.m_brakesTemperature)}, "
            f"Tyres Surface Temperature: {str(self.m_tyresSurfaceTemperature)}, "
            f"Tyres Inner Temperature: {str(self.m_tyresInnerTemperature)}, "
            f"Engine Temperature: {str(self.m_engineTemperature)}, "
            f"Tyres Pressure: {str(self.m_tyresPressure)})"
        )

    def toJSON(self) -> Dict[str, Any]:
        """
        Convert the CarTelemetryData instance to a JSON-compatible dictionary.

        Returns:
            Dict[str, Any]: JSON-compatible dictionary representing the CarTelemetryData instance.
        """

        return {
            "speed": self.m_speed,
            "throttle": self.m_throttle,
            "steer": self.m_steer,
            "brake": self.m_brake,
            "clutch": self.m_clutch,
            "gear": self.m_gear,
            "engine-rpm": self.m_engineRPM,
            "drs": self.m_drs,
            "rev-lights-percent": self.m_revLightsPercent,
            "brakes-temperature": self.m_brakesTemperature,
            "tyres-surface-temperature": self.m_tyresSurfaceTemperature,
            "tyres-inner-temperature": self.m_tyresInnerTemperature,
            "engine-temperature": self.m_engineTemperature,
            "tyres-pressure": self.m_tyresPressure,
            "surface-type": self.m_surfaceType,
        }

    def __eq__(self, other: "CarTelemetryData") -> bool:
        """
        Checks if two CarTelemetryData objects are equal.

        Args:
            other (CarTelemetryData): The other CarTelemetryData object to compare with.

        Returns:
            bool: True if the objects are equal, False otherwise.
        """

        if not isinstance(other, CarTelemetryData):
            return False
        return (
            self.m_speed == other.m_speed
            and self.m_throttle == other.m_throttle
            and self.m_steer == other.m_steer
            and self.m_brake == other.m_brake
            and self.m_clutch == other.m_clutch
            and self.m_gear == other.m_gear
            and self.m_engineRPM == other.m_engineRPM
            and self.m_drs == other.m_drs
            and self.m_revLightsPercent == other.m_revLightsPercent
            and self.m_revLightsBitValue == other.m_revLightsBitValue
            and self.m_brakesTemperature == other.m_brakesTemperature
            and self.m_tyresSurfaceTemperature == other.m_tyresSurfaceTemperature
            and self.m_tyresInnerTemperature == other.m_tyresInnerTemperature
            and self.m_engineTemperature == other.m_engineTemperature
            and self.m_tyresPressure == other.m_tyresPressure
            and self.m_surfaceType == other.m_surfaceType
        )

    def __ne__(self, other: "CarTelemetryData") -> bool:
        """
        Checks if two CarTelemetryData objects are not equal.

        Args:
            other (CarTelemetryData): The other CarTelemetryData object to compare with.

        Returns:
            bool: True if the objects are not equal, False otherwise.
        """
        return not self.__eq__(other)

    def to_bytes(self) -> bytes:
        """
        Convert the CarTelemetryData object to bytes.

        Returns:
            bytes: Bytes representation of the CarTelemetryData object.
        """
        return self.COMPILED_PACKET_STRUCT.pack(
            self.m_speed,
            self.m_throttle,
            self.m_steer,
            self.m_brake,
            self.m_clutch,
            self.m_gear,
            self.m_engineRPM,
            self.m_drs,
            self.m_revLightsPercent,
            self.m_revLightsBitValue,
            self.m_brakesTemperature[0],
            self.m_brakesTemperature[1],
            self.m_brakesTemperature[2],
            self.m_brakesTemperature[3],
            self.m_tyresSurfaceTemperature[0],
            self.m_tyresSurfaceTemperature[1],
            self.m_tyresSurfaceTemperature[2],
            self.m_tyresSurfaceTemperature[3],
            self.m_tyresInnerTemperature[0],
            self.m_tyresInnerTemperature[1],
            self.m_tyresInnerTemperature[2],
            self.m_tyresInnerTemperature[3],
            self.m_engineTemperature,
            self.m_tyresPressure[0],
            self.m_tyresPressure[1],
            self.m_tyresPressure[2],
            self.m_tyresPressure[3],
            self.m_surfaceType[0],
            self.m_surfaceType[1],
            self.m_surfaceType[2],
            self.m_surfaceType[3]
        )

    @classmethod
    def from_values(cls,
        speed: int,
        throttle: int,
        steer: int,
        brake: int,
        clutch: int,
        gear: int,
        engine_rpm: int,
        drs: bool,
        rev_lights_percent: int,
        rev_lights_bit_value: int,
        brakes_temperature_0: int,
        brakes_temperature_1: int,
        brakes_temperature_2: int,
        brakes_temperature_3: int,
        tyres_surface_temperature_0: int,
        tyres_surface_temperature_1: int,
        tyres_surface_temperature_2: int,
        tyres_surface_temperature_3: int,
        tyres_inner_temperature_0: int,
        tyres_inner_temperature_1: int,
        tyres_inner_temperature_2: int,
        tyres_inner_temperature_3: int,
        engine_temperature: int,
        tyres_pressure_0: float,
        tyres_pressure_1: float,
        tyres_pressure_2: float,
        tyres_pressure_3: float,
        surface_type_0: int,
        surface_type_1: int,
        surface_type_2: int,
        surface_type_3: int) -> "CarTelemetryData":
        """
        Create a new CarTelemetryData object from the provided values.

        Args:
            Too many to document for a test method

        Returns:
            CarTelemetryData: A new CarTelemetryData object with the provided values.
        """
        return cls(cls.COMPILED_PACKET_STRUCT.pack(
            speed,
            throttle,
            steer,
            brake,
            clutch,
            gear,
            engine_rpm,
            drs,
            rev_lights_percent,
            rev_lights_bit_value,
            brakes_temperature_0,
            brakes_temperature_1,
            brakes_temperature_2,
            brakes_temperature_3,
            tyres_surface_temperature_0,
            tyres_surface_temperature_1,
            tyres_surface_temperature_2,
            tyres_surface_temperature_3,
            tyres_inner_temperature_0,
            tyres_inner_temperature_1,
            tyres_inner_temperature_2,
            tyres_inner_temperature_3,
            engine_temperature,
            tyres_pressure_0,
            tyres_pressure_1,
            tyres_pressure_2,
            tyres_pressure_3,
            surface_type_0,
            surface_type_1,
            surface_type_2,
            surface_type_3
        ))

class PacketCarTelemetryData(F1PacketBase):
    """
    A class representing telemetry data for multiple cars in a racing simulation.

    Attributes:
        m_header (PacketHeader): Header information for the telemetry data.
        m_carTelemetryData (List[CarTelemetryData]): List of CarTelemetryData objects for each car.
        m_mfdPanelIndex (int): Index of MFD (Multi-Function Display) panel open.
            - 255: MFD closed (Single player, race)
            - 0: Car setup
            - 1: Pits
            - 2: Damage
            - 3: Engine
            - 4: Temperatures
            May vary depending on the game mode.
        m_mfdPanelIndexSecondaryPlayer (int): Secondary player's MFD panel index. See m_mfdPanelIndex for details.
        m_suggestedGear (int): Suggested gear for the player (1-8), 0 if no gear is suggested.
    """

    MAX_CARS = 22
    COMPILED_PACKET_FORMAT_EXTRA = struct.Struct("<BBb")
    PACKET_LEN_EXTRA = COMPILED_PACKET_FORMAT_EXTRA.size

    __slots__ = (
        "m_carTelemetryData",
        "m_mfdPanelIndex",
        "m_mfdPanelIndexSecondaryPlayer",
        "m_suggestedGear"
    )

    def __init__(self, header:PacketHeader, packet: bytes) -> None:
        """
        Initializes a PacketCarTelemetryData object by unpacking the provided binary data.

        Parameters:
            header (PacketHeader): Header information for the telemetry data.
            packet (bytes): Binary data to be unpacked.

        Raises:
            struct.error: If the binary data does not match the expected format.
        """

        super().__init__(header)
        self.m_carTelemetryData: List[CarTelemetryData]

        self.m_carTelemetryData, offset_so_far = _validate_parse_fixed_segments(
            data=packet,
            offset=0,
            item_cls=CarTelemetryData,
            item_len=CarTelemetryData.PACKET_LEN,
            count=self.MAX_CARS,
            max_count=self.MAX_CARS
        )

        self.m_mfdPanelIndex, self.m_mfdPanelIndexSecondaryPlayer, self.m_suggestedGear = \
                self.COMPILED_PACKET_FORMAT_EXTRA.unpack(packet[offset_so_far:])

    def __str__(self) -> str:
        """
        Returns a string representation of the PacketCarTelemetryData object.

        Returns:
            str: String representation of the object.
        """

        telemetry_data_str = ", ".join(str(telemetry) for telemetry in self.m_carTelemetryData)
        mfd_panel_index_to_string_dict = {
            255: "MFD closed Single player, race",
            0: "Car setup",
            1: "Pits",
            2: "Damage",
            3: "Engine",
            4: "Temperatures",
        }
        mfd_panel_string = mfd_panel_index_to_string_dict.get(self.m_mfdPanelIndex, str(self.m_mfdPanelIndex))
        mfd_panel_string_secondary = mfd_panel_index_to_string_dict.get(self.m_mfdPanelIndexSecondaryPlayer,
                                                                        str(self.m_mfdPanelIndexSecondaryPlayer))
        return (
            f"PacketCarTelemetryData("
            f"Header: {str(self.m_header)}, "
            f"Car Telemetry Data: [{telemetry_data_str}], "
            f"MFD Panel Index: {mfd_panel_string}), "
            f"MFD Panel Index Secondary: {mfd_panel_string_secondary}), "
            f"Suggested Gear: {self.m_suggestedGear}"
        )

    def toJSON(self, include_header: bool=False) -> Dict[str, Any]:
        """
        Convert the PacketCarTelemetryData instance to a JSON-compatible dictionary.

        Arguments:
            - include_header - Whether the header dump must be included in the JSON

        Returns:
            Dict[str, Any]: JSON-compatible dictionary representing the PacketCarTelemetryData instance.
        """

        json_data = {
            "car-telemetry-data": [telemetry.toJSON() for telemetry in self.m_carTelemetryData],
            "mfd-panel-index": self.m_mfdPanelIndex,
            "mfd-panel-index-secondary": self.m_mfdPanelIndexSecondaryPlayer,
            "suggested-gear": self.m_suggestedGear
        }
        if include_header:
            json_data["header"] = self.m_header.toJSON()
        return json_data

    def __eq__(self, other: "PacketCarTelemetryData") -> bool:
        """
        Checks if two PacketCarTelemetryData objects are equal.

        Parameters:
            other (PacketCarTelemetryData): The other PacketCarTelemetryData object.

        Returns:
            bool: True if the objects are equal, False otherwise.
        """

        if not isinstance(other, PacketCarTelemetryData):
            return False

        return (
            self.m_header == other.m_header and
            self.m_carTelemetryData == other.m_carTelemetryData and
            self.m_mfdPanelIndex == other.m_mfdPanelIndex and
            self.m_mfdPanelIndexSecondaryPlayer == other.m_mfdPanelIndexSecondaryPlayer and
            self.m_suggestedGear == other.m_suggestedGear
        )

    def __ne__(self, other: "PacketCarTelemetryData") -> bool:
        """
        Checks if two PacketCarTelemetryData objects are not equal.

        Parameters:
            other (PacketCarTelemetryData): The other PacketCarTelemetryData object.

        Returns:
            bool: True if the objects are not equal, False otherwise.
        """

        return not self.__eq__(other)

    def to_bytes(self) -> bytes:
        """
        Convert the PacketCarTelemetryData instance to bytes.

        Returns:
            bytes: Bytes representation of the PacketCarTelemetryData instance.
        """

        raw_bytes = self.m_header.to_bytes()
        raw_bytes += b''.join([telemetry.to_bytes() for telemetry in self.m_carTelemetryData])
        raw_bytes += struct.pack("<BBb",
                                 self.m_mfdPanelIndex, self.m_mfdPanelIndexSecondaryPlayer, self.m_suggestedGear)
        return raw_bytes

    @classmethod
    def from_values(cls,
                     header: PacketHeader,
                     car_telemetry_data: List[CarTelemetryData],
                     mfd_panel_index: int,
                     mfd_panel_index_secondary_player: int,
                     suggested_gear: int) -> "PacketCarTelemetryData":
        """
        Create a PacketCarTelemetryData object from values.

        Args:
            header (PacketHeader): The header of the telemetry packet.
            car_telemetry_data (List[CarTelemetryData]): List of CarTelemetryData objects.
            mfd_panel_index (int): Index of the MFD panel.
            mfd_panel_index_secondary_player (int): Index of the MFD panel for the secondary player.
            suggested_gear (int): Suggested gear for the car.

        Returns:
            PacketCarTelemetryData: A new PacketCarTelemetryData object.
        """

        raw_bytes = b''.join([telemetry.to_bytes() for telemetry in car_telemetry_data])
        raw_bytes += struct.pack("<BBb",
                                 mfd_panel_index, mfd_panel_index_secondary_player, suggested_gear)
        return cls(header, raw_bytes)

File: apps/backend/rust/crates/f1-types/src/packet_6_car_telemetry_data.rs
use crate::PacketHeader;
use crate::errors::InvalidPacketLengthError;
use serde::Serialize;
use serde::ser::{SerializeMap, Serializer};
use std::fmt;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CarTelemetryData {
    pub speed: u16,
    pub throttle: f32,
    pub steer: f32,
    pub brake: f32,
    pub clutch: u8,
    pub gear: i8,
    pub engine_rpm: u16,
    pub drs: bool,
    pub rev_lights_percent: u8,
    pub rev_lights_bit_value: u16,
    pub brakes_temperature: [u16; 4],
    pub tyres_surface_temperature: [u8; 4],
    pub tyres_inner_temperature: [u8; 4],
    pub engine_temperature: u16,
    pub tyres_pressure: [f32; 4],
    pub surface_type: [u8; 4],
}

impl CarTelemetryData {
    pub const PACKET_LEN: usize = 60;

    #[allow(clippy::too_many_arguments)]
    pub fn from_values(
        speed: u16,
        throttle: f32,
        steer: f32,
        brake: f32,
        clutch: u8,
        gear: i8,
        engine_rpm: u16,
        drs: bool,
        rev_lights_percent: u8,
        rev_lights_bit_value: u16,
        brakes_temperature: [u16; 4],
        tyres_surface_temperature: [u8; 4],
        tyres_inner_temperature: [u8; 4],
        engine_temperature: u16,
        tyres_pressure: [f32; 4],
        surface_type: [u8; 4],
    ) -> Self {
        Self {
            speed,
            throttle,
            steer,
            brake,
            clutch,
            gear,
            engine_rpm,
            drs,
            rev_lights_percent,
            rev_lights_bit_value,
            brakes_temperature,
            tyres_surface_temperature,
            tyres_inner_temperature,
            engine_temperature,
            tyres_pressure,
            surface_type,
        }
    }

    pub fn parse(data: &[u8]) -> Result<Self, InvalidPacketLengthError> {
        if data.len() != Self::PACKET_LEN {
            return Err(InvalidPacketLengthError::new(format!(
                "Received packet length {} expected {}",
                data.len(),
                Self::PACKET_LEN
            )));
        }

        let mut offset = 0usize;
        let read_u8 = |data: &[u8], offset: &mut usize| {
            let value = data[*offset];
            *offset += 1;
            value
        };
        let read_i8 = |data: &[u8], offset: &mut usize| {
            let value = data[*offset] as i8;
            *offset += 1;
            value
        };
        let read_u16 = |data: &[u8], offset: &mut usize| {
            let value = u16::from_le_bytes(
                data[*offset..*offset + 2]
                    .try_into()
                    .expect("fixed slice length"),
            );
            *offset += 2;
            value
        };
        let read_f32 = |data: &[u8], offset: &mut usize| {
            let value = f32::from_le_bytes(
                data[*offset..*offset + 4]
                    .try_into()
                    .expect("fixed slice length"),
            );
            *offset += 4;
            value
        };

        Ok(Self {
            speed: read_u16(data, &mut offset),
            throttle: read_f32(data, &mut offset),
            steer: read_f32(data, &mut offset),
            brake: read_f32(data, &mut offset),
            clutch: read_u8(data, &mut offset),
            gear: read_i8(data, &mut offset),
            engine_rpm: read_u16(data, &mut offset),
            drs: read_u8(data, &mut offset) != 0,
            rev_lights_percent: read_u8(data, &mut offset),
            rev_lights_bit_value: read_u16(data, &mut offset),
            brakes_temperature: [
                read_u16(data, &mut offset),
                read_u16(data, &mut offset),
                read_u16(data, &mut offset),
                read_u16(data, &mut offset),
            ],
            tyres_surface_temperature: [
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
            ],
            tyres_inner_temperature: [
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
            ],
            engine_temperature: read_u16(data, &mut offset),
            tyres_pressure: [
                read_f32(data, &mut offset),
                read_f32(data, &mut offset),
                read_f32(data, &mut offset),
                read_f32(data, &mut offset),
            ],
            surface_type: [
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
                read_u8(data, &mut offset),
            ],
        })
    }

    pub fn to_bytes(&self) -> [u8; Self::PACKET_LEN] {
        let mut bytes = [0u8; Self::PACKET_LEN];
        let mut offset = 0usize;
        let write_u8 = |bytes: &mut [u8], offset: &mut usize, value: u8| {
            bytes[*offset] = value;
            *offset += 1;
        };
        let write_i8 = |bytes: &mut [u8], offset: &mut usize, value: i8| {
            bytes[*offset] = value as u8;
            *offset += 1;
        };
        let write_u16 = |bytes: &mut [u8], offset: &mut usize, value: u16| {
            bytes[*offset..*offset + 2].copy_from_slice(&value.to_le_bytes());
            *offset += 2;
        };
        let write_f32 = |bytes: &mut [u8], offset: &mut usize, value: f32| {
            bytes[*offset..*offset + 4].copy_from_slice(&value.to_le_bytes());
            *offset += 4;
        };

        write_u16(&mut bytes, &mut offset, self.speed);
        write_f32(&mut bytes, &mut offset, self.throttle);
        write_f32(&mut bytes, &mut offset, self.steer);
        write_f32(&mut bytes, &mut offset, self.brake);
        write_u8(&mut bytes, &mut offset, self.clutch);
        write_i8(&mut bytes, &mut offset, self.gear);
        write_u16(&mut bytes, &mut offset, self.engine_rpm);
        write_u8(&mut bytes, &mut offset, self.drs as u8);
        write_u8(&mut bytes, &mut offset, self.rev_lights_percent);
        write_u16(&mut bytes, &mut offset, self.rev_lights_bit_value);
        for value in self.brakes_temperature {
            write_u16(&mut bytes, &mut offset, value);
        }
        for value in self.tyres_surface_temperature {
            write_u8(&mut bytes, &mut offset, value);
        }
        for value in self.tyres_inner_temperature {
            write_u8(&mut bytes, &mut offset, value);
        }
        write_u16(&mut bytes, &mut offset, self.engine_temperature);
        for value in self.tyres_pressure {
            write_f32(&mut bytes, &mut offset, value);
        }
        for value in self.surface_type {
            write_u8(&mut bytes, &mut offset, value);
        }
        bytes
    }
}

impl fmt::Display for CarTelemetryData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "CarTelemetryData(Speed: {}, Throttle: {}, Steer: {}, Brake: {}, Clutch: {}, Gear: {}, Engine RPM: {}, DRS: {}, Rev Lights Percent: {}, Brakes Temperature: {:?}, Tyres Surface Temperature: {:?}, Tyres Inner Temperature: {:?}, Engine Temperature: {}, Tyres Pressure: {:?})",
            self.speed,
            self.throttle,
            self.steer,
            self.brake,
            self.clutch,
            self.gear,
            self.engine_rpm,
            self.drs,
            self.rev_lights_percent,
            self.brakes_temperature,
            self.tyres_surface_temperature,
            self.tyres_inner_temperature,
            self.engine_temperature,
            self.tyres_pressure
        )
    }
}

impl Serialize for CarTelemetryData {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(15))?;
        map.serialize_entry("speed", &self.speed)?;
        map.serialize_entry("throttle", &self.throttle)?;
        map.serialize_entry("steer", &self.steer)?;
        map.serialize_entry("brake", &self.brake)?;
        map.serialize_entry("clutch", &self.clutch)?;
        map.serialize_entry("gear", &self.gear)?;
        map.serialize_entry("engine-rpm", &self.engine_rpm)?;
        map.serialize_entry("drs", &self.drs)?;
        map.serialize_entry("rev-lights-percent", &self.rev_lights_percent)?;
        map.serialize_entry("brakes-temperature", &self.brakes_temperature)?;
        map.serialize_entry("tyres-surface-temperature", &self.tyres_surface_temperature)?;
        map.serialize_entry("tyres-inner-temperature", &self.tyres_inner_temperature)?;
        map.serialize_entry("engine-temperature", &self.engine_temperature)?;
        map.serialize_entry("tyres-pressure", &self.tyres_pressure)?;
        map.serialize_entry("surface-type", &self.surface_type)?;
        map.end()
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct PacketCarTelemetryData {
    pub header: PacketHeader,
    pub car_telemetry_data: Vec<CarTelemetryData>,
    pub mfd_panel_index: u8,
    pub mfd_panel_index_secondary_player: u8,
    pub suggested_gear: i8,
}

impl PacketCarTelemetryData {
    pub const MAX_CARS: usize = 22;
    pub const PACKET_LEN_EXTRA: usize = 3;
    pub const PAYLOAD_LEN: usize =
        (Self::MAX_CARS * CarTelemetryData::PACKET_LEN) + Self::PACKET_LEN_EXTRA;

    pub fn parse(header: PacketHeader, packet: &[u8]) -> Result<Self, InvalidPacketLengthError> {
        if packet.len() != Self::PAYLOAD_LEN {
            return Err(InvalidPacketLengthError::new(format!(
                "Received packet length {} is not equal to expected {}",
                packet.len(),
                Self::PAYLOAD_LEN
            )));
        }

        let telemetry_bytes = Self::MAX_CARS * CarTelemetryData::PACKET_LEN;
        let car_telemetry_data = packet[..telemetry_bytes]
            .chunks_exact(CarTelemetryData::PACKET_LEN)
            .map(CarTelemetryData::parse)
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            header,
            car_telemetry_data,
            mfd_panel_index: packet[telemetry_bytes],
            mfd_panel_index_secondary_player: packet[telemetry_bytes + 1],
            suggested_gear: packet[telemetry_bytes + 2] as i8,
        })
    }

    pub fn from_values(
        header: PacketHeader,
        car_telemetry_data: Vec<CarTelemetryData>,
        mfd_panel_index: u8,
        mfd_panel_index_secondary_player: u8,
        suggested_gear: i8,
    ) -> Self {
        Self {
            header,
            car_telemetry_data,
            mfd_panel_index,
            mfd_panel_index_secondary_player,
            suggested_gear,
        }
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(PacketHeader::PACKET_LEN + Self::PAYLOAD_LEN);
        bytes.extend_from_slice(&self.header.to_bytes());
        for telemetry in &self.car_telemetry_data {
            bytes.extend_from_slice(&telemetry.to_bytes());
        }
        bytes.push(self.mfd_panel_index);
        bytes.push(self.mfd_panel_index_secondary_player);
        bytes.push(self.suggested_gear as u8);
        bytes
    }
}

impl fmt::Display for PacketCarTelemetryData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let telemetry = self
            .car_telemetry_data
            .iter()
            .map(ToString::to_string)
            .collect::<Vec<_>>()
            .join(", ");
        write!(
            f,
            "PacketCarTelemetryData(Header: {}, Car Telemetry Data: [{}], MFD Panel Index: {}, MFD Panel Index Secondary: {}, Suggested Gear: {})",
            self.header,
            telemetry,
            self.mfd_panel_index,
            self.mfd_panel_index_secondary_player,
            self.suggested_gear
        )
    }
}

impl Serialize for PacketCarTelemetryData {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(4))?;
        map.serialize_entry("car-telemetry-data", &self.car_telemetry_data)?;
        map.serialize_entry("mfd-panel-index", &self.mfd_panel_index)?;
        map.serialize_entry(
            "mfd-panel-index-secondary",
            &self.mfd_panel_index_secondary_player,
        )?;
        map.serialize_entry("suggested-gear", &self.suggested_gear)?;
        map.end()
    }
}
